language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Ck.hs

```diff
@@ -90,45 +202,50 @@ substituteDb varFor new = go where
 -- > s ▷ unwrap M   ↦ s , (unwrap _)   ▷ M
 -- > s ▷ abs α K M  ↦ s ◁ abs α K M
 -- > s ▷ lam x A M  ↦ s ◁ lam x A M
+-- > s ▷ builtin bn ↦ s ◁ builtin bn (arity bn) (arity bn) [] []
 -- > s ▷ con cn     ↦ s ◁ con cn
 -- > s ▷ error A    ↦ ◆
 (|>)
     :: (GShow uni, GEq uni, DefaultUni <: uni)
-    => Context uni -> Plain Term uni -> CkM uni (Plain Term uni)
-stack |> TyInst _ fun ty      = FrameTyInstArg ty  : stack |> fun
-stack |> Apply _ fun arg      = FrameApplyArg arg  : stack |> fun
-stack |> IWrap _ pat arg term = FrameIWrap pat arg : stack |> term
-stack |> Unwrap _ term        = FrameUnwrap        : stack |> term
-stack |> tyAbs@TyAbs{}        = stack <| tyAbs
-stack |> lamAbs@LamAbs{}      = stack <| lamAbs
-stack |> bi@Builtin{}         = stack <| bi
-stack |> constant@Constant{}  = stack <| constant
-_     |> err@Error{}          =
-    throwingWithCause _EvaluationError (UserEvaluationError ()) $ Just err
-_     |> var@Var{}            =
-    throwingWithCause _MachineError OpenTermEvaluatedMachineError $ Just var
+    => Context uni -> Term TyName Name uni () -> CkM uni (Term TyName Name uni ())
+stack |> TyInst  _ fun ty        = FrameTyInstArg ty  : stack |> fun
+stack |> Apply   _ fun arg       = FrameApplyArg arg  : stack |> fun
+stack |> IWrap   _ pat arg term  = FrameIWrap pat arg : stack |> term
+stack |> Unwrap  _ term          = FrameUnwrap        : stack |> term
+stack |> TyAbs   _ tn k term     = stack <| VTyAbs tn k term
+stack |> LamAbs  _ name ty body  = stack <| VLamAbs name ty body
+stack |> Builtin _ bn            = do
+                                    arity <- arityOf bn
+                                    stack <| VBuiltin bn arity arity [] []
+stack |> c@Constant{}          = stack <| VCon c
+_     |> _err@Error{}          =
+    throwingWithCause _EvaluationError (UserEvaluationError CkEvaluationFailure) $ Nothing
+_     |> _var@Var{}            =
+    throwingWithCause _MachineError OpenTermEvaluatedMachineError $ Nothing
+
 
+-- FIXME: make sure that the specification is up to date and that this matches.
 -- | The returning part of the CK machine. Rules are as follows:
 --
--- > s , {_ A}           ◁ abs α K M  ↦ s         ▷ M
+-- > s , {_ A}           ◁ abs α K M  ↦ s         ▷ {A/α}M
 -- > s , [_ N]           ◁ V          ↦ s , [V _] ▷ N
 -- > s , [(lam x A M) _] ◁ V          ↦ s         ▷ [V/x]M
--- > s , {_ A}           ◁ F          ↦ s ◁ {F A}  -- Partially saturated constant.
--- > s , [F _]           ◁ V          ↦ s ◁ [F V]  -- Partially saturated constant.
+-- > s , {_ A}           ◁ F          ↦ s ◁ {F A}  -- Partially saturated constant. ???
+-- > s , [F _]           ◁ V          ↦ s ◁ [F V]  -- Partially saturated constant. ???
```

Comment:
Thanks.  I think some of this might be a bit out of date now, but I'll check and/or fix that after we've written this stuff down properly in the PLC specification.